src/Service/Bulk/BulkCreateUsersServiceConsoleProxy.php

```diff
@@ -0,0 +1,127 @@
+use OAT\SimpleRoster\Lti\Service\StateDrivenUserGenerator;
+use OAT\SimpleRoster\Repository\AssignmentRepository;
+use OAT\SimpleRoster\Repository\Criteria\FindLineItemCriteria;
+use OAT\SimpleRoster\Repository\LineItemRepository;
+use OAT\SimpleRoster\Repository\LtiInstanceRepository;
+use OAT\SimpleRoster\Lti\Service\AssigmentFactoryInterface;
+use OAT\SimpleRoster\Storage\UserGenerator\StorageInterface;
+use OAT\SimpleRoster\Entity\LineItem;
+
+class BulkCreateUsersServiceConsoleProxy
+{
+    private LineItemRepository $lineItemRepository;
+    private GenerateGroupIdsService $generateGroupIdsService;
+    private LtiInstanceRepository $ltiInstanceRepository;
+    private UserCreationResultMessage $userCreationMessage;
+    private BulkCreateUsersService $service;
+
+    public function __construct(
+        LineItemRepository $lineItemRepository,
+        GenerateGroupIdsService $generateGroupIdsService,
+        LtiInstanceRepository $ltiInstanceRepository,
+        UserCreationResultMessage $userCreationMessage,
+        BulkCreateUsersService $service
+    ) {
+        $this->lineItemRepository = $lineItemRepository;
+        $this->generateGroupIdsService = $generateGroupIdsService;
+        $this->userCreationMessage = $userCreationMessage;
+        $this->ltiInstanceRepository = $ltiInstanceRepository;
+        $this->service = $service;
+    }
+
+    /**
+     * @throws LineItemNotFoundException
+     */
+    public function createUsers(
+        array $lineItemIds,
+        array $lineItemSlugs,
+        array $userPrefixes,
+        int $batchSize,
+        ?string $groupPrefix
+    ): UserCreationResult {
+        $notExistLineItemsArray = [];
+
+        $lineItems = [];
+        if (!empty($lineItemIds) || !empty($lineItemSlugs)) {
+            $criteria = (new FindLineItemCriteria())
+                ->addLineItemIds(...$lineItemIds)
+                ->addLineItemSlugs(...$lineItemSlugs);
+            $lineItems = $this->lineItemRepository->findLineItemsByCriteria($criteria)->jsonSerialize();
+
+            if (empty($lineItems)) {
+                $exceptionMessage = $lineItemIds
+                    ? implode(',', $lineItemIds) . ' Line item id(s) not exist in the system'
+                    : implode(',', $lineItemSlugs) . ' Line item slug(s) not exist in the system';
+                throw new LineItemNotFoundException($exceptionMessage);
+            }
+
+            $notExistLineItemsArray = $lineItemIds
+                ? array_diff($lineItemIds, array_map(fn($item) => $item->getId(), $lineItems))
+                : array_diff($lineItemSlugs, array_map(fn($item) => $item->getSlug(), $lineItems));
+        }
+
+        if (empty($lineItemIds) && empty($lineItemSlugs)) {
+            $lineItems = $this->lineItemRepository->findAllAsCollection()->jsonSerialize();
+            if (empty($lineItems)) {
+                throw new LineItemNotFoundException('No line items were found in database.');
+            }
+        }
```

Comment:
I believe that we could rework a little bit this part of the code to get rid of this nested 'if' conditions and make the code more readable. For example this way :)
```
        $criteria = new FindLineItemCriteria();
        if (!empty($lineItemIds)) {
            $criteria->addLineItemIds(...$lineItemIds);
            $exceptionMessage = implode(',', $lineItemIds) . ' Line item id(s) not exist in the system';
        }

        if (!empty($lineItemSlugs)) {
            $criteria->addLineItemSlugs(...$lineItemSlugs);
            $exceptionMessage = implode(',', $lineItemSlugs) . ' Line item slug(s) not exist in the system';
        }

        $criteria = $criteria->hasLineItemIdsCriteria() || $criteria->hasLineItemSlugsCriteria() ? $criteria : null;

        $lineItems = $criteria
            ? $this->lineItemRepository->findLineItemsByCriteria($criteria)->jsonSerialize()
            : $this->lineItemRepository->findAllAsCollection()->jsonSerialize();

        if (empty($lineItems)) {
            throw new LineItemNotFoundException($exceptionMessage ?? 'No line items were found in database.');
        }

        $notExistLineItemsArray = [];
        if ($criteria) {
            $notExistLineItemsArray = $lineItemIds
                ? array_diff($lineItemIds, array_map(fn($item) => $item->getId(), $lineItems))
                : array_diff($lineItemSlugs, array_map(fn($item) => $item->getSlug(), $lineItems));
        }
```

Reply:
You'r right. It's already kinda refactored part :). I think in the next iteration we'll remove this completely. But for now it's  not part of the task. PR is too huge.